File: 1367 Linked List in Binary Tree/1367.cpp
class Solution {
public:
    bool valid(ListNode* h, TreeNode* p){
        return !h || p && p->val == h->val && (valid(h->next, p->left) || valid(h->next, p->right));
    }

    bool dfs(ListNode* h, TreeNode* p){
        return p && (valid(h, p) or dfs(h, p->left) or dfs(h, p->right));
    }

    bool isSubPath(ListNode* head, TreeNode* root) {
        return dfs(head, root);
    }
};

File: 1367 Linked List in Binary Tree/1367.py
class Solution:
    def isSubPath(self, h: Optional[ListNode], root: Optional[TreeNode]) -> bool:
        def valid(p, q):
            return not q or p and p.val == q.val and (valid(p.left, q.next) or valid(p.right, q.next))

        def dfs(p):
            return p and (valid(p, h) or dfs(p.left) or dfs(p.right))

        return dfs(root)
